src/submodule.rs

```diff
@@ -33,6 +33,21 @@ impl<'repo> Submodule<'repo> {
         unsafe { crate::opt_bytes(self, raw::git_submodule_branch(self.raw)) }
     }
 
+    /// Perform the clone step for a newly created submodule.
+    ///
+    /// This performs the necessary `git_clone` to setup a newly-created submodule.
+    pub fn clone(&mut self, opts: &mut SubmoduleUpdateOptions<'_>) -> Result<Repository, Error> {
```

Comment:
I think `opts` should probably be an `Option`, similar to how [`update`](https://github.com/rust-lang/git2-rs/blob/c2a26133731d152f67a4fb11b35c4e139441e60c/src/submodule.rs#L186-L201) is done.

Reply:
According to the docs the values isn't optional though: https://libgit2.org/libgit2/#v1.0.1/group/submodule/git_submodule_clone.

What we could do is create a default instance if the Option value is `None`. If that is Ok with you.

Reply:
The libgit2 docs aren't consistent or necessarily complete, and in this case I think it is fine to pass NULL.  The code [handles NULL](https://github.com/libgit2/libgit2/blob/c71321a099373753c22055921c8f538afed5ebb6/src/submodule.c#L845), and I think it would be unlikely that they would remove that.

Another example, [`git_rebase_open`](https://github.com/libgit2/libgit2/blob/c71321a099373753c22055921c8f538afed5ebb6/include/git2/rebase.h#L210) doesn't mention NULL is supported, but it is, and the rust bindings [can pass NULL](https://github.com/rust-lang/git2-rs/blob/c2a26133731d152f67a4fb11b35c4e139441e60c/src/repo.rs#L1987).

Reply:
I did the change but that's where I hit my knowledge limitations. I can't get it to compile properly. The whole pointer stuff doesn't work out. I do know why (because `*mut git_submodule_opts` is not the same as the type of `ptr::null()`) but I don't know how to fix this :(

Reply:
The issue here is that `SubmoduleUpdateOptions::raw` returns a *value* and not a pointer.  `ptr::null` is a pointer, so the types aren't the same.  The call to `git_submodule_clone` needs a pointer, so the solution is to place the value in a (temporary) local variable, and pass a pointer to it, something like this:

```rust
        unsafe {
            let mut raw_repo = ptr::null_mut();
            let raw_opts = opts.map(|o| o.raw());
            try_call!(raw::git_submodule_clone(
                &mut raw_repo,
                self.raw,
                raw_opts.as_ref()
            ));
            Ok(Binding::from_raw(raw_repo))
        }
```

The call to `as_ref` turns it to an `Option<&git_submodule_update_options>`, which is a nullable pointer. `Option<T>` has a guarantee that None will translate to NULL for ffi bindings.

I would suggest to also update the unit test to use None instead of default, just to double check that. 😄

Reply:
> The call to `as_ref` turns it to an `Option<&git_submodule_update_options>`, which is a nullable pointer. `Option<T>` (for certain types of T) has a guarantee that None will translate to NULL for ffi bindings.

Does this hold true for all ffi bindings or is that something that has been setup in git2?

Reply:
> Does this hold true for all ffi bindings or is that something that has been setup in git2?

This is mostly macro magic in the git2 library.  The [`try_call`](https://github.com/rust-lang/git2-rs/blob/c2a26133731d152f67a4fb11b35c4e139441e60c/src/call.rs#L12-L19) does a bunch of implicit conversions (in this case converting `Option<&git_submodule_update_options>` to `*const git_submodule_update_options`), which handles the conversion of None to null (around [here](https://github.com/rust-lang/git2-rs/blob/c2a26133731d152f67a4fb11b35c4e139441e60c/src/call.rs#L96-L100) I think).

There is a guarantee in some cases that Option can represent a null (see [nullable pointer optimization](https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization) for more detail), but I believe this particular scenario is not leveraging that.